### PR TITLE
Add new vuln endpoint and new parameters

### LINKED ISSUE #8863: Update the API endpoins to provide the detection timestamp, severity, and score of the CVEs as well as the scans timestamps
|Wazuh version|Component|
|---|---|
| 4.3 | API |

As part of the epic https://github.com/wazuh/wazuh/issues/8727, we are adding more information to the inventory of vulnerabilities that affect each agent. Having said that, there are some API endpoints that should be updated in order to provide the necessary information to be used in the Wazuh UI.

- The `GET /agents/{agent_id}/cve` endpoint should now include the detection timestamp, severity, and score of the CVE. This information was added to the `vuln_cves` table of the agents' databases in https://github.com/wazuh/wazuh/issues/8735. So the information should be taken from the `detection_time`, `severity`, `CVSS2`, and `CVSS3` columns.
- The information related to the scans timestamps should be provided in an API endpoint. This information is already in the agents' databases in the table `vuln_metadata`. The columns `LAST_FULL_SCAN` and `LAST_PARTIAL_SCAN` contains the timestamp of the last full and partial scan performed over the agent respectively.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

### PR DESCRIPTION
**Pull request by Selutario**
|Related issue|
|---|
| Closes #8863 |

## Description

Hello team!

This PR adds the following new fields in the API response to `GET /vulnerability/{agent_id}` endpoint:
- cvss2_score
- cvss3_score
- detection_time
- severity

However, only severity has been added as a filter on the endpoint. The rest of the new fields, as they are numeric, make more sense to be used to obtain ranges that can be specified within the q parameter, such as: 

> `GET /vulnerability/001?q=cvss3_score>9&severity=critical`
```JSON
{
  "data": {
    "affected_items": [
      {
        "architecture": "amd64",
        "severity": "Critical",
        "version": "3.8.5-1~20.04",
        "name": "python3.8-minimal",
        "cvss2_score": 7.5,
        "cve": "CVE-2021-3177",
        "cvss3_score": 9.8,
        "type": "PACKAGE",
        "detection_time": "2021-06-14T11:25:46Z",
        "status": "VALID"
      }
    ],
    "total_affected_items": 67,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All selected vulnerabilities were returned",
  "error": 0
}
```

Regards,
Selu.

### CODE REVIEW

File: framework/wazuh/core/vulnerability.py [hunk 1]
```diff
@@ -5,23 +5,40 @@
 from wazuh.core import common
 from wazuh.core.agent import Agent
 from wazuh.core.utils import WazuhDBQuery, WazuhDBBackend
+from datetime import datetime
```
Comment (AdriiiPRodri) [inline framework/wazuh/core/vulnerability.py:8]: Imports are not sorted

File: framework/wazuh/vulnerability.py [hunk 2]
```diff
@@ -9,8 +9,6 @@
 from wazuh.core.results import AffectedItemsWazuhResult
 from wazuh.rbac.decorators import expose_resources
 
-logger = logging.getLogger('wazuh')
```
Comment (AdriiiPRodri) [inline framework/wazuh/vulnerability.py:12]: Unused import `logging`

File: framework/wazuh/core/vulnerability.py [hunk 3]
```diff
@@ -5,23 +5,40 @@
...
+
+    def _format_data_into_dictionary(self):
+        def format_fields(field_name, value):
+            if field_name in ['detection_time', 'last_partial_scan', 'last_full_scan']:
+                try:
+                    return datetime.utcfromtimestamp(int(value))
+                except (ValueError, TypeError):
+                    return value
+            else:
+                return value
```
Comment (AdriiiPRodri) [inline framework/wazuh/core/vulnerability.py:40]:
```suggestion
                    pass

            return value
```

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

File: framework/wazuh/vulnerability.py (hunk 2 above)
Comment (Selutario) [reply framework/wazuh/vulnerability.py:12]: It was removed

**Review by AdriiiPRodri — APPROVED** [review]
LGTM!

File: framework/wazuh/core/vulnerability.py [hunk 4]
```diff
@@ -11,17 +13,33 @@ class WazuhDBQueryVulnerability(WazuhDBQuery):
...
         backend = WazuhDBBackend(agent_id)
-        WazuhDBQuery.__init__(self, offset=offset, limit=limit, table='vuln_cves', sort=sort, search=search,
+        WazuhDBQuery.__init__(self, offset=offset, limit=limit, table=table, sort=sort, search=search,
                               select=select, fields=fields, default_sort_field=default_sort_field,
                               default_sort_order='ASC', filters=filters, query=query, backend=backend,
                               min_select_fields=set(), count=count, get_data=get_data, distinct=distinct)
+
+    def _format_data_into_dictionary(self):
+        def format_fields(field_name, value):
+            if field_name in ['detection_time', 'last_partial_scan', 'last_full_scan']:
```
Comment (davidjiglesias) [inline framework/wazuh/core/vulnerability.py:35]: Isnt it better to define a variable (maybe class one) with all date fields?

File: framework/wazuh/vulnerability.py [hunk 5]
```diff
@@ -57,3 +53,32 @@ def get_agent_cve(agent_list=None, offset=0, limit=common.database_limit, sort=N
     result.total_affected_items = data['totalItems']
 
     return result
+
+
+@expose_resources(actions=["vulnerability:read"], resources=["agent:id:{agent_list}"], post_proc_func=None)
+def last_scan(agent_list):
+    """Return when the last full and partial vulnerability scan of a specified agent end.
```
Comment (davidjiglesias) [inline framework/wazuh/vulnerability.py:60]:
```suggestion
    """Return when the last full and partial vulnerability scan of a specified agent ended.
```

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**